ikarus/python/dirichletvalues/dirichletvalues.hh

```diff
@@ -20,22 +20,80 @@
 #include <dune/python/pybind11/stl_bind.h>
 
 #include <ikarus/finiteelements/ferequirements.hh>
+
 // PYBIND11_MAKE_OPAQUE(std::vector<bool>);
 namespace Ikarus::Python {
 
+template <class DirichletValues>
+void forwardCorrectFunction(DirichletValues& dirichletValues, const pybind11::function& functor, auto&& cppFunction) {
+  using Basis        = typename DirichletValues::Basis;
+  using Intersection = typename Basis::GridView::Intersection;
+  using BackendType  = typename DirichletValues::BackendType;
+  using MultiIndex   = typename Basis::MultiIndex;
+
+  using LocalViewWrapper = Dune::Python::LocalViewWrapper<Basis>;
+
+  using FixBoundaryDOFsWithGlobalIndexFunction = std::function<void(Eigen::Ref<Eigen::VectorX<bool>>, int)>;
+  using FixBoundaryDOFsWithLocalViewFunction =
+      std::function<void(Eigen::Ref<Eigen::VectorX<bool>>, int, LocalViewWrapper&)>;
+  using FixBoundaryDOFsWithIntersectionFunction =
+      std::function<void(Eigen::Ref<Eigen::VectorX<bool>>, int, LocalViewWrapper&, const Intersection&)>;
+
+  // Disambiguate by number of arguments, as casting doesn't properly work with functions
+  pybind11::module inspect_module = pybind11::module::import("inspect");
+  pybind11::object result         = inspect_module.attr("signature")(functor).attr("parameters");
+  size_t numParams                = pybind11::len(result);
+
+  if (numParams == 2) {
+    auto function = functor.template cast<const FixBoundaryDOFsWithGlobalIndexFunction>();
+    auto lambda   = [&](BackendType& vec, const MultiIndex& indexGlobal) {
+      // we explicitly only allow flat indices
+      function(vec.vector(), indexGlobal[0]);
+    };
+    cppFunction(lambda);
+
+  } else if (numParams == 3) {
+    auto function = functor.template cast<const FixBoundaryDOFsWithLocalViewFunction>();
+    auto lambda   = [&](BackendType& vec, int localIndex, auto& lv) {
+      auto lvWrapper = LocalViewWrapper(lv.rootLocalView());
```

Comment:
Why do you construct the `rootLocalView` instead of the one from the subBasis?

Reply:
The `LocalViewWrapper` seems to only be able to constructed from a `LocalView`, but in this case `lv` is a `SubSpaceLocalView` whose interface is a bit different, thats why i resorted to `rootLocalView`. 

If you think this is an issue I can invetigate a bit further

Reply:
So for now this interface is strange then. You pass in a subbasis but you get the root local view. 
The problem is that type you define `LocalViewWrapper` templated with the root basis.
If you just write inside the function 
```cpp
using LocalView = std::remove_cvref_t<decltype(lv)>;
using LocalViewWrapper =  Dune::Python::LocalViewWrapper<typename LocalView::GlobalBasis>;
auto lvWrapper= LocalViewWrapper(lv);
using FixBoundaryDOFsWithLocalViewFunction =
      std::function<void(Eigen::Ref<Eigen::VectorX<bool>>, int, LocalViewWrapper&)>;

```

Reply:
This results in a compiler error. 
As I have stated, the `LocalViewWrapper` is not compatible with a `SubSpaceLocalView`, so we have to make a PR to dune-functions to make a `SubSpaceLocalViewWrapper` or we implement one in Ikarus itself

Reply:
Ok but then does this mean in Python:

```py
basis ...

sb =subSpaceBasis(basis)
lv = sb.localView()
```
also fails?

Reply:
Okay mhm that seems to work and return a 
```
<class 'dune.generated.subspaceBasis_21d9edbc1071a405709a7b58e97f7ff1.LocalView'>
```

Reply:
Okay sorry, I was wrong, it is possible to create a LocalViewWrapper from a SubSpaceBasis, there is still a compiler error, it might have to do with the cpp implementation
```cpp
else if constexpr (Concepts::IsFunctorWithArgs<F, BackendType, int, typename Basis::LocalView,
                                                     typename Basis::GridView::Intersection>) {
                                                     ```
as we are here using Basis::LocalView

Reply:
Ah yes thats wrong for this case

Reply:
So this works as expected at the Cpp side, but on the python side i get the following runtime error
```
RuntimeError: Unable to convert call argument '2' of type 'Dune::Python::LocalViewWrapper<Dune::Functions::SubspaceBasis<Dune::Functions::DefaultGlobalBasis<Dune::Functions::PowerPreBasis<Dune::Functions::BasisFactory::FlatLexicographic, Dune::Functions::LagrangePreBasis<Dune::GridView<Dune::DefaultLeafGridViewTraits<Dune::YaspGrid<2, Dune::EquidistantOffsetCoordinates<double, 2> > const> >, 1, double>, 2ul> >, Dune::TypeTree::HybridTreePath<> > >' to Python object
```

Reply:
I also tried registering the correct LocalViewWrapper on the fly like this 
https://github.com/ikarus-project/ikarus/blob/8c9855881b09e82450bd5bd8437ea87943a09352/ikarus/python/dirichletvalues/dirichletvalues.hh#L86
But unfortunatly i then get the following error 
```
ValueError: couldn't find requested type Dune::Functions::SubspaceBasis<Dune::Functions::DefaultGlobalBasis<Dune::Functions::PowerPreBasis<Dune::Functions::BasisFactory::FlatLexicographic, Dune::Functions::LagrangePreBasis<Dune::GridView<Dune::DefaultLeafGridViewTraits<Dune::YaspGrid<2, Dune::EquidistantOffsetCoordinates<double, 2> > const> >, 1, double>, 2ul> >, Dune::TypeTree::HybridTreePath<> > in type registry
```
Maybe if you have time to have a look. Ive commited my not-working work in progress on that

Reply:
You have to also register the subspace basis: Like in https://gitlab.dune-project.org/staging/dune-functions/-/blame/master/dune/python/functions/subspacebasis.hh?ref_type=heads#L27-28.
Basically the mechanism is always like this: https://github.com/ikarus-project/ikarus/blob/1f6d41cd844adcdd36c8dc7c76da1d7a93894baa/ikarus/python/finiteelements/registerferequirements.hh#L25-L44
So insert the class and if it is already registerd do nothing and other wise register all the member functions.

```cpp
  typedef Dune::Python::LocalViewWrapper<FlatBasis> LocalViewWrapper;
  auto includes = Dune::Python::IncludeFiles{"dune/python/functions/globalbasis.hh"};
  auto [lv, isNotRegistered] = Dune::Python::insertClass<LocalViewWrapper>(
                scopedf, "LocalViewWrapper",
                Dune::Python::GenerateTypeName("Dune::Python::LocalViewWrapper", Dune::MetaType<FlatBasis>()),
                includes)
           
if(isNotRegistered)
{  lv.def("bind", &LocalViewWrapper::bind);
  lv.def("unbind", &LocalViewWrapper::unbind);
  lv.def("index", [](const LocalViewWrapper& localView, int index) { return localView.index(index); });
  lv.def("__len__", [](LocalViewWrapper& self) -> int { return self.size(); });

  Dune::Python::Functions::registerTree<typename LocalViewWrapper::Tree>(lv);
  lv.def("tree", [](const LocalViewWrapper& view) { return view.tree(); });

}
```
Can you also fix this in https://github.com/ikarus-project/ikarus/blob/8c9855881b09e82450bd5bd8437ea87943a09352/ikarus/python/finiteelements/fe.hh#L117C1-L130C76 .

Similar can be done for the subSpace basis. But for the subSpace basis there is actually a trick:

You can directly try to register the subspacebasis in Python by calling simply:


https://gitlab.dune-project.org/staging/dune-functions/-/blame/master/python/dune/functions/subspacebasis.py?ref_type=heads#L14

at 
https://github.com/ikarus-project/ikarus/blob/8c9855881b09e82450bd5bd8437ea87943a09352/python/ikarus/dirichlet_values.py#L14 .

This registers then the correct subspaceBasis and inside it will work then.

Reply:
How do i get the basis object in this context ?

> https://github.com/ikarus-project/ikarus/blob/8c9855881b09e82450bd5bd8437ea87943a09352/python/ikarus/dirichlet_values.py#L14
> .
> 
> This registers then the correct subspaceBasis and inside it will work then.

Reply:
Is this still relevant?

Reply:
I don't think so